### ISSUE TITLE
Create JSON file and related functionality to support replacement of hard-coded issue labels

### ISSUE THREAD

**Issue by rdhmdhl** [hackforla/website#5205]
### Overview

- Labels are hard-coded and spread throughout our Github Action files. 
- If a label is changed in the repo, we have to investigate where that label was being used in Github Actions, and update it.
- We need to add a JSON file containing key/value pairs of the labels in the repository, and import it into each GHA file that uses labels

### Action Items

- [x] Create JSON file with key/value pairs for all repository labels (e.g. a "label directory"). This file should include:
  - [x]  Create a `Key` for each label, written in camelCase: e.g. key for "Time Sensitive" would be "timeSensititve", etc.
  - [x]  The `Value` should be an array of the ["label id", "label display name"]
- [x]  Save JSON file to repo in a new subfolder called `github-actions/utils/_data/label-directory.json`
- [x] Write JS file that queries the JSON file and returns the label display name when given the Key  

To be useful and meet the requirements that initiated this issue, the JSON cannot be static. Whenever a label is updated with a new name or when a new label is added to the repo, the change must propagate dynamically and update the JSON file (i.e. JSON needs CRUD.)
- [x] Create a GHA that is triggered `on: label: types: [edited, created, deleted]` with an associated JS file to update and/or add to the "label directory"
- [x] The GHA will need to write **and commit** changes to the new "label-directory"   
- [x] Prepare ER to document new GHA and label lookups in wiki #6454  

Merge Team:
- [x] When merged, remove dependency from #6454 and change status to "New Issue Approval"
### Resources/Instructions

- Summary of HfLA GitHub Actions [here](https://github.com/hackforla/website/wiki/HfLA-GitHub-Actions)
- Workflow files in HfLA repository [here](https://github.com/hackforla/website/tree/gh-pages/.github/workflows)
- For creating the JSON file, see JSON object literals documentation [here](https://www.w3schools.com/js/js_json_objects.asp)
- Issue #4050 and [related DR](https://github.com/hackforla/website/issues/4050#issuecomment-1646282976)
- Issue #6389

**Comment by roslynwythe** [comment]
@rdhmdhl In the Action Items, please specify the filepath of the json file.  Also in a comment please describe for us the process we will use to update the file.  When you are done please apply the `ready for dev lead` label.  Thanks!

**Comment by ExperimentsInHonesty** [comment]
@rdhmdhl It looks like you have not touched this issue in months.  I am unassigning it and kicking it back to dev lead for review and reprioritization.

**Comment by t-will-gillis** [comment]
@ExperimentsInHonesty 

## Update 3/11/24: I jumped the gun and I have already worked on this issue. Marking it `Ready for prioritization`- please let me know if you would like any changes, otherwise it is ready to go.


_original message:_
Hi @ExperimentsInHonesty  @roslynwythe I think this issue needs some more info and since this is `time sensitive`, I will be adding that info shortly. 


Unless you disagree or tell me otherwise I will continue with:
-	I know we probably won’t use emojis on all the labels, but it makes sense to me that while we are doing this we would/should create key-value pairs for all labels whether or not we use them
-	The easy way to assemble a list of all of our labels is using the API (there are over 100). Since this is a `time sensitive` issue, I can do this- unless you would like to leave this for whoever works on the issue(?)
-	I think we will want to give guidance about what to name the keys for the JSON or JS object- suggested naming as in [DR for Emojis](https://github.com/hackforla/website/issues/4050#issuecomment-1646282976)
-	Also guidance about what to name the file and where to store it - something like `return-labels.js` in the `/utils/` folder
-	I think that testing would be done for all files, not just `check-labels.js`, and would be handled by the issues initiated by #6389. I would not use Nektos for this. I feel like it is complicated and the test results are not useful for GHAs involving the Project Board. I also admit that I might be wrong and someone else might know something I don’t about Nektos, but until then…I think the testing should be how we have been doing it.

**Comment by t-will-gillis** [comment]
Updates: 
I am going to assign myself to work on this issue. 

An important "Action Item" missing on the description above is that there needs to be a mechanism for updating the JSON file reference whenever we change a label. 
- Reason: whenever we change a label in the future, this change has to occur both on the actual label object & on the JSON file. If we only change the JSON, then with each change a brand new label is going to be created automatically in the repo- this would become untenable after a short time. If we only change the Label itself via "Edit labels", this change will not be reflected on the JSON when a GHA runs.  
- Solution: we can add a GHA triggered by the label change. The GHA would then update the JSON automatically to make sure it reflects the current label. In this case then, once we create the JSON we don't want to be editing it ourselves ~(except for the case of adding a completely new label.)~ GHA to add new label to JSON

**Comment by t-will-gillis** [comment]
Info / to dos for self: 
- [x] [github_context](https://github.com/t-will-gillis/website/actions/runs/8179451675/job/22365551859)
- [x] extract id of what triggered GHA
- [x] use id to update the display name in the JSON
- [x] save the data (`stefanzweifel/git-auto-commit-action`?)
- [x] Note that if Id does not exist in JSON, can be added now
- [x] Google Sheets 'label directory '- in order to create/prepare an initial JSON
- [x] [Emoji Cheat Sheet](https://github.com/ikatyang/emoji-cheat-sheet)
- [x] ~confirm use of variables in Markdown~ 
- [ ] Need to talk about possible ways to address the labels used in the `/ISSUE_TEMPLATE/` Markdown files
    - Directly/ manually updates templates each time we change a label
    - Write a linux script in a GHA to search the templates and update to the correct label each time a change is made (could do something like this for all of the label references and ditch the directory)
    - Remove the labels from all of the templates, then use `issue-trigger.yml`, `check-labels.js`, etc. to apply the correct labels to the appropriate issue template
    - something else?
- [ ] Wiki!
- [ ] Do we need some kind of periodic confirmation that the **_repo's_** (not Google Sheets') "label directory" has not drifted/ accumulated errors <-- this is probably overkill

**Comment by t-will-gillis** [comment]
eta: 3/10/24

**Comment by ExperimentsInHonesty** [comment]
@t-will-gillis we actually have a spreadsheet and a script that has all the lables and pulls new ones in.  We just need to add to it the variable information.  Its current use is for supporting the dashboard.

**Comment by ExperimentsInHonesty** [comment]
I guess the point I am making is we need one source of truth.

**Comment by t-will-gillis** [comment]
@ExperimentsInHonesty I will add this to the Monday meeting schedule. 

Where is the spreadsheet and script? Is it [this one?](https://docs.google.com/spreadsheets/d/1-ltg0qMeZSgOnqrCU0nKUDQd1JOXTMWrNTK63VZjXdk/edit#gid=0)


For what it is worth, this is how I was thinking about it (apologies for talking through it- helps me to understand):  

We know that:  
- Currently when the GHAs refer to a label, it is a reference to the actual label name value (eg `2 weeks inactive`).
- We are creating a new JSON file to match up new 'key' labels to the existing 'value' labels. ("2WeeksInactive" : `2 weeks inactive`)
- Now instead of the actual name value of the label, the GHAs reference the key value, and we have internal function to read the JSON and return the actual value for the key. 

So far so good. Next, my thought was what happens when we want to change a label, for example if the final label is going to be `🥱 2 weeks inactive`. One way to do it:
- Add a completely new label to the repo, then manually edit the JSON to make sure that the old key now refers to the new name value  (while making sure that there are no differences in white space or letter casing or the GHA won't find the label). The old label still exists, and now there also is a new label to keep track of. For the dashboard I assume that both labels would need to be tracked now since both labels refer to `2 weeks inactive`  

What I am thinking:
- If we want the label to be `🥱 2 weeks inactive`, we just edit the `2 weeks inactive` label directly. This can trigger a GHA that updates the JSON file automatically to the new value for the same key. Then since we are editing the existing label itself, this new label gets applied retroactively to previous issues- it is still the same label. 
- The source of truth would be the repo itself.
- For the Dashboard: if the script pulls the labels via the label Id, nothing needs to change because the Id is the same. If the Dashboard script pulls the labels via the name, we will need to update the script, but that would be no worse than the previous option...

**Comment by ExperimentsInHonesty** [comment]
https://github.com/hackforla/live-dashboard-automation
https://github.com/hackforla/website/issues/5813

**Comment by t-will-gillis** [comment]
- The Dashboard uses the [Labels for Weekly Label Check](https://docs.google.com/spreadsheets/d/1-ltg0qMeZSgOnqrCU0nKUDQd1JOXTMWrNTK63VZjXdk/edit#gid=0) worksheet for the labels, however:
  - This worksheet is not complete and in its current state it can't be used for what we need. 
  - It is missing many required labels from our repo such as `2 weeks inactive`, `To Update !`, `Status: Updated` and includes labels our repo does not have `Complexity: 8pt`, etc. 
- There is an upcoming issue, #5810, for `role: data analyst`, and mentions using the GitHub API to retrieve all existing repo labels then ultimately saving these to an empty spreadsheet on the Drive. 
- When 5810 becomes available, we can use this as a starting point or reference back to it. Or if the [Labels for Weekly Label Check](https://docs.google.com/spreadsheets/d/1-ltg0qMeZSgOnqrCU0nKUDQd1JOXTMWrNTK63VZjXdk/edit#gid=0) worksheet was modified we could use it to gather all of our repos labels.   

How do we want to move forward with this issue? 



- From previous [Label Directory](https://docs.google.com/spreadsheets/d/1XkX2Q1dA91LSzzyp1ueTHnEZW55EEHriKBhiILJFLlc/edit#gid=0) - all of our current labels with keys. 
- This [JSON](https://github.com/t-will-gillis/website/blob/create-json-label-directory-5205/github-actions/utils/_data/label-directory.json) was built from the Label Directory worksheet. Helper functions exist to update the JSON whenever the labels themselves are updated.
- The current GHA is able to query the current JSON Label Directory as we need it to.

**Comment by ExperimentsInHonesty** [comment]
@t-will-gillis I have added this to the next agenda (either the 13th or the 20th) depending on when I can attend.  But I would be happy to discuss it sooner if you're available.  I have sent you a Slack message about making an appt with me.

**Comment by ExperimentsInHonesty** [comment]
- after #6454 is done, then we will need to removed this from the icebox https://github.com/hackforla/website/issues/5810#issuecomment-2099305953

**Comment by t-will-gillis** [comment]
Progress: 
@ExperimentsInHonesty I believe this works as intended, other than tweaks and final location of the Source of Truth file (Source of Truth in Website Drive currently). Do we want to review it one last time?

- Created "Current Labels- Website Repo" Google Sheets file:
  - Apps Script queries GitHub and records all of the Website's current "labelNames" and "labelIds". 
  - After first use, this Apps Script and worksheet will query the GH repo labels on a schedule so that this info can be compared with the 'Source of Truth' document. (For first use, creates temporary "keyNames" so that a "Label Object" can be saved as JSON)
- Created `label-directory.json` JSON file to store reference info on "Label Object"
- Created `update-label-directory.yml` workflow, triggered by GitHub Label events: edits, deletions, and creations
  - These label events are compared to the JSON file. For label edits and creations, JSON "Label Object" is updated. For label deletions, "labelId" is reset to '9999999999' but "Label Object" is not removed from JSON.
  - Writes the label events to an "Updates Log" on the 'Source of Truth' file. 
- Created 'Source of Truth'  - "Label Object" updated in 'Source of Truth' with flag notification of change.
- This is a [working copy of the 'Source of Truth'](https://docs.google.com/spreadsheets/d/1Y1-ba1ITSs6L6Av0mtX0Qy-Oq-tsKP_rwR-AJfuDYCk/edit) document- the four entries are test examples showing created, edited, and deleted labels.
- The GHA workflow and Google Apps Script work together to keep the actual GitHub labels, the JSON, and the 'Source of Truth' coordinated.
- Code is in place so that GitHub Actions can reference the "keyName" of the "Label Object" and return with a "Label Name"
